### ISSUE TITLE
truncate after select may cause distributed deadlock in mx nodes

### ISSUE THREAD

**Issue by mtuncer** [citusdata/citus#2373]
Observed a distributed deadlock when testing a recent work on allowing truncate on MX nodes.

Verified that the deadlock does not occur on single node (not distributed) configuration.

- create 2 node cluster with both workers as mx nodes
- create a distributed table table_1

- connect to worker 2 (the one added to mx later)  (connection 1)
  -  start a transaction
  -  issue select * from table_1

- connect to worker 1 (connection 2)
  - start a transaction
  - issue truncate table table_1 --> will block

- switch to connection 1
  - issue truncate table table_1 --> will block

After a while our distribution deadlock detection kicks in an cancel one of the transactions.

**Comment by onderkalaci** [comment]
adding some notes:

- This seems not a likely case that users would run, given that it includes concurrent `TRUNCATE`s from worker nodes, and one of the transactions that the `TRUNCATE` is executed includes some other command.
- There seems no other commands except for `TRUNCATE` to lead such issue
- The reason for this happening is that we execute `TRUNCATE` in the `AFTER` trigger. This means that we cannot enforce strict ordering on the nodes to execute `lock_relation_if_exists` or such UDFs because some nodes would have already executed `TRUNCATE` locally before others lock it.
- To me it seems OK to leave as a known issue and not fix it. But, if we're going to expand the use of `lock_relation_if_exists`, we should consider this case.

**Comment by marcocitus** [comment]
This no longer reproduces.

**Comment by onderkalaci** [comment]
> This no longer reproduces.

I can reproduce the issue. If we are to rely on `lock_relation_if_exists` trick for more objects from workers, we should make sure to understand the reason and resolve this

> The reason for this happening is that we execute TRUNCATE in the AFTER trigger. This means that we cannot enforce strict ordering on the nodes to execute lock_relation_if_exists or such UDFs because some nodes would have already executed TRUNCATE locally before others lock it.

And, this sounds wrong.

**Comment by onderkalaci** [comment]
Can we repro this on PG?

**Comment by hanefi** [comment]
> Can we repro this on PG?

No, we can't. In the last step, truncates are not blocked on connection 1, while connection 2 waits for acquiring the lock.
